ChameleonMiniGUI/Json/MifareUltralightModel.cs

```diff
@@ -37,32 +37,148 @@ public MifareUltralightCardInfo Card
+                    .Concat(stba(info.Pack))
+                    .Concat(stba(info.TBO_1))
+                    .Concat(stba(info.Signature))
+                    .Concat(Blocks.SelectMany(bytes => bytes))
+                    .ToArray();
+            else
+                // Save as simple dump format
+                return Blocks.SelectMany(bytes => bytes)
+                    .ToArray();
+        }
+
+        public static bool HasUltralightNewHeader(IReadOnlyList<byte> bytes)
+        {
+            if (bytes.Count % 4 != 0 || bytes.Count <= MifareUltralightCardInfo.NewPrefixLength)
+                return false;
+
+            // tbo should be ZERO
+            if (bytes[8] != 0x00 || bytes[9] != 0x00)
+                return false;
+
+            // tbo1 should be ZERO
+            if (bytes[10] != 0x00)
+                return false;
+
+            // pages count must be equals to pages in header
+            int maxPage = (bytes.Count - MifareUltralightCardInfo.NewPrefixLength) / 4 - 1;
+            return maxPage == bytes[11];
+        }
+
+        public static bool HasUltralightHeader(IReadOnlyList<byte> bytes)
+        {
+            // empty header
+            var empty_header = new byte[]
+            {
+                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
+            };
+
+            if (bytes.SequenceEqual(empty_header))
+            {
+                return true;
+            }
+
+            // detect mfu header. If probability of magic values is more than 50%, assume file has header.
+            var probability = 0d;
+            // first two bytes of version should be 0x00, 0x04
+            if (bytes[0] == 0x00 && bytes[1] == 0x04)
+                probability += 0.25;
+
+            // tbo should be ZERO
+            if (bytes[8] == 0x00 && bytes[9] == 0x00)
+                probability += 0.15;
+
+            // tbo1 should be ZERO
+            if (bytes[15] == 0x00)
+                probability += 0.15;
+
+            // tearing is normally 0xBD
+            if (bytes[10] == 0xBD || bytes[11] == 0xBD || bytes[12] == 0xBD)
+                probability += 0.35;
+
+            return (probability >= 0.50);
         }
 
         public static MifareUltralightModel Parse(byte[] data)
         {
+            bool isNewHeaderFormat = HasUltralightNewHeader(data);
+            bool isOldHeaderFormat = !isNewHeaderFormat && HasUltralightHeader(data);
+            int headerLength = isNewHeaderFormat ? MifareUltralightCardInfo.NewPrefixLength : isOldHeaderFormat ? MifareUltralightCardInfo.PrefixLength : 0;
+            int pages = (data.Length - headerLength) / 4;
             return new MifareUltralightModel()
             {
                 Created = "ChameleonMiniGUI",
-                Card = new MifareUltralightCardInfo()
+                Card = isNewHeaderFormat ?
+                new MifareUltralightCardInfo()
```

Comment:
Make this return of object nice and clean.  Not nested like this.

Reply:
fixed